cdc/sink/mq.go

```diff
@@ -0,0 +1,167 @@
+	partitionNum int32
+
+	sinkCheckpointTsCh chan uint64
+	globalResolvedTs   uint64
+	checkpointTs       uint64
+}
+
+func newMqSink(mqProducer mqProducer.Producer) *mqSink {
+	partitionNum := mqProducer.GetPartitionNum()
+	return &mqSink{
+		mqProducer:         mqProducer,
+		partitionNum:       partitionNum,
+		sinkCheckpointTsCh: make(chan uint64, 128),
+	}
+}
+
+func (k *mqSink) EmitResolvedEvent(ctx context.Context, ts uint64) error {
+	atomic.StoreUint64(&k.globalResolvedTs, ts)
+	return nil
+}
+
+func (k *mqSink) EmitCheckpointEvent(ctx context.Context, ts uint64) error {
+	keyByte, err := model.NewResolvedMessage(ts).Encode()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = k.mqProducer.BroadcastMessage(ctx, keyByte, nil)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (k *mqSink) EmitRowChangedEvent(ctx context.Context, rows ...*model.RowChangedEvent) error {
+	var sinkCheckpointTs uint64
+	for _, row := range rows {
+		if row.Resolved {
+			sinkCheckpointTs = row.Ts
+			continue
+		}
+		partition := k.calPartition(row)
+		key, value := row.ToMqMessage()
+		keyByte, err := key.Encode()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		valueByte, err := value.Encode()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		err = k.mqProducer.SendMessage(ctx, keyByte, valueByte, partition)
+		if err != nil {
+			log.Error("send message failed", zap.ByteStrings("row", [][]byte{keyByte, valueByte}), zap.Int32("partition", partition))
+			return errors.Trace(err)
+		}
+	}
+	if sinkCheckpointTs == 0 {
+		return nil
+	}
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case k.sinkCheckpointTsCh <- sinkCheckpointTs:
+	}
+	return nil
+}
+
+func (k *mqSink) calPartition(row *model.RowChangedEvent) int32 {
+	hash := crc32.NewIEEE()
+	_, err := hash.Write([]byte(row.Schema))
+	if err != nil {
+		log.Fatal("calculate hash of message key failed, please report a bug", zap.Error(err))
+	}
+	_, err = hash.Write([]byte(row.Table))
+	if err != nil {
+		log.Fatal("calculate hash of message key failed, please report a bug", zap.Error(err))
+	}
+
+	return int32(hash.Sum32() % uint32(k.partitionNum))
+}
+
+func (k *mqSink) EmitDDLEvent(ctx context.Context, ddl *model.DDLEvent) error {
+	key, value := ddl.ToMqMessage()
+	keyByte, err := key.Encode()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	valueByte, err := value.Encode()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	err = k.mqProducer.BroadcastMessage(ctx, keyByte, valueByte)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+func (k *mqSink) CheckpointTs() uint64 {
+	return atomic.LoadUint64(&k.checkpointTs)
+}
+
+func (k *mqSink) Run(ctx context.Context) error {
+	for {
+		var sinkCheckpointTs uint64
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case sinkCheckpointTs = <-k.sinkCheckpointTsCh:
+		}
+		globalResolvedTs := atomic.LoadUint64(&k.globalResolvedTs)
+		if globalResolvedTs < sinkCheckpointTs {
```

Comment:
In which case will `globalResolvedTs` be less than `sinkCheckpointTs`

Reply:
https://github.com/pingcap/ticdc/pull/322/files#diff-172dd021d09a9fac1ffc69e5792d4c5bR83

sinkCheckpointTs is localResolvedTs in the processor, when we add a table, old local resolvedTS(min resolvedTs of old tables) may be greater than resolvedTS of the new table. At this moment, local resolvedTS will be fallback.

when local resolvedTS is fallback, we will postpone to pushing global resolvedts:
https://github.com/pingcap/ticdc/blob/master/cdc/processor.go#L526

Reply:
Oh I see. Also adding some comments here.

Reply:
added